android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -361,6 +370,60 @@ class TranslationsMiddleware(
         }
     }
 
+    /**
+     * Retrieves the list of languages and their settings and dispatches the result to the
+     * [BrowserState.translationEngine] via [TranslationsAction.SetLanguageSettingsAction] or
+     * else dispatches the failure.
+     *
+     * For failure dispatching:
+     * If a tab ID is not provided, then only [TranslationsAction.EngineExceptionAction] will be
+     * dispatched to set the error on the [BrowserState.translationEngine].
+     *
+     * If a tab ID is provided, then  [TranslationsAction.EngineExceptionAction]
+     * AND [TranslationsAction.TranslateExceptionAction] will be dispatched
+     * to set the error both on the [BrowserState.translationEngine] and
+     * [SessionState.translationsState].
+     *
+     * @param context Context to use to dispatch to the store.
+     * @param tabId If a Tab ID is associated with the request for error handling.
+     * If null, this will only dispatch errors on the global translations browser state.
+     *
+     */
+    private fun requestLanguageSettings(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        tabId: String? = null,
+    ) {
+        engine.getLanguageSettings(
+
+            onSuccess = { settings ->
+                context.store.dispatch(
+                    TranslationsAction.SetLanguageSettingsAction(
+                        languageSettings = settings,
+                    ),
+                )
+                logger.info("Success requesting language settings.")
+            },
+
+            onError = {
+                context.store.dispatch(
+                    TranslationsAction.EngineExceptionAction(
+                        error = TranslationError.CouldNotLoadLanguageSettingsError(it),
+                    ),
+                )
+
+                if (tabId != null) {
+                    TranslationsAction.TranslateExceptionAction(
+                        tabId = tabId,
+                        operation = TranslationOperation.FETCH_AUTOMATIC_LANGUAGE_SETTINGS,
+                        translationError = TranslationError.CouldNotLoadLanguageSettingsError(it),
+                    )
```

Comment:
>I am kinda getting lost if not dizzy from just the many engines and operations we have.

I agree, I've been wondering about how best to manage error state too. There is something that definitely can be improved, but am not yet sure about the best way to go about it. 

Some of what is tricky is there is a lot of error handling, both globally and per-tab. I'm not sure the best way to consolidate and also keep everything in sync. The pattern I'm trying right now is `Init or global request = global error`, `tab request = session + global error`. A failed language request is also going to impact `Tab 2`, even if `Tab 1` requested. (But, I don't necessarily want to overwrite `Tab 2`s error.) I don't expect a tab to be calling the languages, unless something went wrong during `Init` too.

A good start to the refactor is `translationError`'s name should ideally be refactored to something like `generalError` or  `error`, `sessionError`. (I'm not totally sure `settingsError` is necessary, but  at the time was it seemed like it was a different path/non-core error and didn't want to overwrite a core error, which is another consideration.) I have a feeling there might be a nicer pattern to avoid overwriting errors too. 

One major concern outside of the intertwinedness/complexity I have right now is wondering how overwriting errors is going to impact error handling. (e.g., a. more recent error can overwrite an older error). The solution might be an error map that it is tied to operation or something like `TranslationOperation`  keys and `TranslationError` values. (There would be some redundancy, since implicitly certain errors are tied to certain operations, but it might be clearer in the views what to check for and since this has grown so much 😵‍💫.)

There is a lot to explore here, I'll file a bug to investigate!